meilisearch-core/src/query_builder.rs

```diff
@@ -93,36 +76,66 @@ impl<'c, 'f, 'd> QueryBuilder<'c, 'f, 'd> {
         query: &str,
         range: Range<usize>,
     ) -> MResult<(Vec<Document>, usize)> {
+        let facets_docids: Option<sdset::SetBuf<DocumentId>> = if let Some(facets) = self.facets {
+            let mut ands = Vec::new();
+            for f in facets {
+                match f {
+                    Either::Left(ors) => {
+                        let mut sets = Vec::new();
+                        for key in ors {
+                            if let Some(docids) = self.index.facets.document_ids(reader, key)? {
+                                // can we do better than that?
+                                sets.push(sdset::SetBuf::new_unchecked(docids.into_owned()));
+                            }
+                        }
+                        let sets = sets.iter().map(|s| s.as_set()).collect::<Vec<_>>();
+                        ands.push(sdset::multi::OpBuilder::from_vec(sets).union().into_set_buf());
+                    }
+                    Either::Right(key) =>{
+                        if let Some(docids) = self.index.facets.document_ids(reader, key)? {
+                            ands.push(sdset::SetBuf::new_unchecked(docids.into_owned()));
+                        }
+                    }
+                };
+            }
+            let ands = ands.iter().map(sdset::SetBuf::as_ref).collect::<Vec<_>>();
+            Some(sdset::multi::OpBuilder::from_vec(ands).intersection().into_set_buf())
+        } else {
+            None
+        };
```

Comment:
I am not sure this is the best implementation of this algorithm, we should thing about that when you got time.